### PR TITLE
RSE-1147:  Update recurring contribution

### PR DESCRIPTION
**Pull request by erawat**
## Overview
This PR is to add functionalities to update, a recurring contribution record, selected contribution record, create a direct debit mandate and attach a created mandate to existing contribution record. 

## Before
The functionality does not exist. 

## After

After the form is submitted: 
- Recurring Contribution payment method is updated to 'direct_debit'
- Cycle day is set to next payment run date from today's date. 
- Create a new  mandate 
- Attach a created mandate to existing recurring contribution and a selected contribution.
- Confirmation page is shown after the form is submitted.

### CODE REVIEW

File: CRM/ManualDirectDebit/Form/SetUp.php [hunk 1]
```diff
@@ -70,6 +72,68 @@ public function buildQuickForm() {
 
   public function postProcess() {
     parent::postProcess();
+
+    $values = $this->exportValues();
+
+    $contributionRecur = civicrm_api3('Contribution', 'getsingle', [
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Form/SetUp.php:78]: I try to use a rule of thumb to try and keep methods under 50 lines of code, I think this one is getting up close to that amount. Might be a good idea to refactor these API calls into ither methods to improve readibility and separate concerns. This one in particular got me confused, as the variable's name is `$contributionRecur`, but the call is to 'Contribution' entity initially. So I though, why did he call the variable that way? But then I realized there was a linked API call to obtain the recurring contribution.

Eg:
```php
public function postProcess() {
  parent::postProcess();

  $values = $this->exportValues();

  $contributionRecur = $this->getRecurringContribution($values['contribution_id']);
  $this->updateRecurringContribution($contributionRecur, $values);
  $this->createDirectDebitMandate($contributionRecur, $values);

  $url = CRM_Utils_System::url('civicrm/direct_debit/setup/confirmation');
  CRM_Utils_System::redirect( $url);
}
```

File: CRM/ManualDirectDebit/Form/SetUp.php [hunk 2]
```diff
@@ -70,6 +72,68 @@ public function buildQuickForm() {
...
+      'id' => $values['contribution_id'],
+      'api.ContributionRecur.get' => [],
+    ])['api.ContributionRecur.get']['values'][0];
+
+    $cycleDay = $this->getCycleDay($values['payment_dates'], $contributionRecur['frequency_unit']);
+
+    $test = civicrm_api3('ContributionRecur', 'create', [
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Form/SetUp.php:85]: `$test` is a bit generic. Plus it doesn't seem to be used anywhere else in the code, so I guess we can just remove it.

File: CRM/ManualDirectDebit/Form/SetUp.php [hunk 3]
```diff
@@ -92,12 +156,12 @@ private function getPaymentDates() {
     $locale = civicrm_api3('Setting', 'get', [
       'sequential' => 1,
       'return' => ["lcMessages"],
-    ])['values']['lcMessages'];
+    ])['values'][0]['lcMessages'];
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Form/SetUp.php:159]: Wow, good catch!

File: CRM/ManualDirectDebit/Form/SetUp.php (hunk 2 above)
Comment (erawat) [reply CRM/ManualDirectDebit/Form/SetUp.php:85]: @MiyaNoctem  good catch, I was using it for debugging the result. Will remove.

File: CRM/ManualDirectDebit/Form/SetUp.php (hunk 1 above)
Comment (erawat) [reply CRM/ManualDirectDebit/Form/SetUp.php]: @MiyaNoctem the reason I am doing that because I don't want APIs twice.

**Pull request merged**